### ISSUE TITLE
Are libgfortran_64_3.dll and Fortran the trouble maker on visual studio? 

### ISSUE THREAD

**Issue by yyy-git** [xianyi/OpenBLAS#1437]
Windows 10,x64.OpenBLAS compiled by msys(int git)+mingw(in tdm-gcc).

Everytme the libopenblas.dll works right if I use gcc,but it troubles me to call it inside visual studio project.I have tried cl /DYNAMICBASE:NO or make DYNAMICARCH=1, and neither can work in all situations.

Now I'm tired of setting the options of proect again and again,so I use LoadLibrary() and GetProcAddress() instead.

daxpy: blas Fortran-interface.
cblas_daxpy: blas C-interface.
dgesv: lapack Fortran-interface.
LAPACKE_dgesv: lapack C-interface.
D:\LIB\OpenBLAS-v0.2.14-Win64-int32\bin\libopenblas.dll: downloaded from the Internet, the readme says 
		make QUIET_MAKE=1 TARGET=NEHALEM DYNAMIC_ARCH=1 HOSTCC=gcc NUM_THREADS=64 BINARY=64 CC=x86_64-w64-mingw32-gcc FC=x86_64-w64-mingw32-gfortran

 .
D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL\bin\libopenblas.dll: make TARGET=HASWELL
D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_CBLAS\lib\libopenblas.dll: make TARGET=HASWELL NO_LAPACK=1  ONLY_CBLAS=1
D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_NOFORTRAN\lib\libopenblas.dll: make TARGET=HASWELL ONLY_CBLAS=1
D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_DYN\lib\libopenblas.dll: make TARGET=HASWELL DYNAMIC_ARCH=1
D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_NOFORTRAN2\lib\libopenblas.dll: make TARGET=HASWELL NOFORTRAN=1 NUM_THREADS=64 BINARY=64

I'm not aware of that ONLY_CBLAS=1 implies NO_LAPACK=1. The result of ...\HASWELL_NOFORTRAN\... is exactly the same as ...\HASWELL_CBLAS\...
Now I find that with ONLY_CBLAS=1, NO_LAPACK=1 or NOFORTRAN=1, the object file libopenblas.dll dosen't depend on serval other dlls, that's the key why it works.
All other libopenblas.dll depends on libgcc_s_seh_64-1.dll,libgfortran_64-3.dll,and libgccquadmath_64-0.dll.

Whenever it fails to load the .dll file, it must be the fault of libgfortran_64_3.dll. Maybe I used  incompatible compile tool chain.

Anyway, what's the point of libgcc_s_seh_64-1.dll, libgfortran_64-3.dll,and libquadmath_64-0.dll in LAPACK? I never see quad precision functions or useful Fortran source files in OpenBLAS...
and what's the relationship between Fortran and lapack? Why can't lapack be compiled without Fortran?

Besides, I extracts the functions in mkl.dll, only Fortran interface found. Some prefer Fortran, but not me...

output from CMD:
Load library:D:\LIB\OpenBLAS-v0.2.14-Win64-int32\bin\libopenblas.dll successed.
function daxpy found.
function dgesv found.
function LAPACKE_dgesv found.
Get Proc:cblas_daxpy successed.
double vector:        1       1       1       1       1       1       1       1       1       1
 + 2.0 *
double vector:        0       1       2       3       4       5       6       7       8       9
 =
double vector:        1       3       5       7       9      11      13      15      17      19

Load library:D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL\bin\libopenblas.dll failed.
Error code:1114.
Error: 动态链接库(DLL)初始化例程失败。
Get func failed.

Load library:D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_CBLAS\lib\libopenblas.dll successed.
Get Proc:cblas_daxpy successed.
double vector:        1       1       1       1       1       1       1       1       1       1
 + 2.0 *
double vector:        0       1       2       3       4       5       6       7       8       9
 =
double vector:        1       3       5       7       9      11      13      15      17      19

Load library:D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_NOFORTRAN\lib\libopenblas.dll successed.
Get Proc:cblas_daxpy successed.
double vector:        1       1       1       1       1       1       1       1       1       1
 + 2.0 *
double vector:        0       1       2       3       4       5       6       7       8       9
 =
double vector:        1       3       5       7       9      11      13      15      17      19

Load library:D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_NOFORTRAN2\lib\libopenblas.dll successed.
function daxpy found.
Get Proc:cblas_daxpy successed.
double vector:        1       1       1       1       1       1       1       1       1       1
 + 2.0 *
double vector:        0       1       2       3       4       5       6       7       8       9
 =
double vector:        1       3       5       7       9      11      13      15      17      19

Load library:D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_DYN\lib\libopenblas.dll failed.
Error code:1114.
Error: 动态链接库(DLL)初始化例程失败。
Get func failed.

Load library:C:\Program Files\MATLAB\R2017a\bin\win64\libmwlapack.dll successed.
function dgesv found.
function LAPACKE_dgesv found.
Get Proc:cblas_daxpy failed.
Error code:127.
Error: 找不到指定的程序。
Get func failed.

Load library:C:\Program Files\MATLAB\R2017a\bin\win64\mkl.dll successed.
function daxpy found.
function dgesv found.
function LAPACKE_dgesv found.
Get Proc:cblas_daxpy failed.
Error code:127.
Error: 找不到指定的程序。
Get func failed.

output from WinDbg(when it works):
ModLoad: 00000000`00400000 00000000`0040b000   image00000000`00400000
ModLoad: 00007fff`6a140000 00007fff`6a320000   ntdll.dll
ModLoad: 00007fff`67720000 00007fff`677ce000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007fff`66740000 00007fff`669a6000   C:\WINDOWS\System32\KERNELBASE.dll
ModLoad: 00007fff`67b30000 00007fff`67bcd000   C:\WINDOWS\System32\msvcrt.dll
ModLoad: 00000000`6d7c0000 00000000`6dbd5000   D:\LIB\OpenBLAS-v0.2.20-Win64\HASWELL_CBLAS\lib\libopenblas.dll
ModLoad: 00007fff`677d0000 00007fff`6795f000   C:\WINDOWS\System32\USER32.dll
ModLoad: 00007fff`66b80000 00007fff`66ba0000   C:\WINDOWS\System32\win32u.dll
ModLoad: 00007fff`69300000 00007fff`69328000   C:\WINDOWS\System32\GDI32.dll
ModLoad: 00007fff`665a0000 00007fff`66733000   C:\WINDOWS\System32\gdi32full.dll
ModLoad: 00007fff`673f0000 00007fff`6748b000   C:\WINDOWS\System32\msvcp_win.dll
ModLoad: 00007fff`672f0000 00007fff`673e6000   C:\WINDOWS\System32\ucrtbase.dll
(1c70.2898): Break instruction exception - code 80000003 (first chance)
ntdll!LdrpDoDebuggerBreak+0x30:
00007fff`6a212ebc cc              int     3

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
0:000> g
ModLoad: 00007fff`69b90000 00007fff`69bbd000   C:\WINDOWS\System32\IMM32.DLL
ModLoad: 00007fff`69bd0000 00007fff`69ed8000   C:\WINDOWS\System32\combase.dll
ModLoad: 00007fff`69780000 00007fff`6989f000   C:\WINDOWS\System32\RPCRT4.dll
ModLoad: 00007fff`67490000 00007fff`67502000   C:\WINDOWS\System32\bcryptPrimitives.dll
ModLoad: 00007fff`66420000 00007fff`6648e000   C:\WINDOWS\System32\dtrampo.dll
ModLoad: 00007fff`67a80000 00007fff`67b21000   C:\WINDOWS\System32\ADVAPI32.dll
ModLoad: 00007fff`69120000 00007fff`6917b000   C:\WINDOWS\System32\sechost.dll

output from WinDbg(when it dosen't work):
Executable search path is: 
ModLoad: 00007ff7`57ef0000 00007ff7`57ef8000   openblas_teste.exe
ModLoad: 00007fff`6a140000 00007fff`6a320000   ntdll.dll
ModLoad: 00007fff`67720000 00007fff`677ce000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007fff`66740000 00007fff`669a6000   C:\WINDOWS\System32\KERNELBASE.dll
ModLoad: 00007fff`672f0000 00007fff`673e6000   C:\WINDOWS\System32\ucrtbase.dll
ModLoad: 00000000`6d7c0000 00000000`6fbe3000   D:\LIB\dlls\libopenblas.dll
ModLoad: 00007fff`67b30000 00007fff`67bcd000   C:\WINDOWS\System32\msvcrt.dll
ModLoad: 00007fff`677d0000 00007fff`6795f000   C:\WINDOWS\System32\USER32.dll
ModLoad: 00007fff`66b80000 00007fff`66ba0000   C:\WINDOWS\System32\win32u.dll
ModLoad: 00007fff`69300000 00007fff`69328000   C:\WINDOWS\System32\GDI32.dll
ModLoad: 00007fff`665a0000 00007fff`66733000   C:\WINDOWS\System32\gdi32full.dll
ModLoad: 00007fff`673f0000 00007fff`6748b000   C:\WINDOWS\System32\msvcp_win.dll
ModLoad: 00007fff`5f220000 00007fff`5f236000   C:\WINDOWS\SYSTEM32\VCRUNTIME140.dll
ModLoad: 00000000`63200000 00000000`63345000   D:\LIB\dlls\libgfortran_64-3.dll
ModLoad: 00007fff`67a80000 00007fff`67b21000   C:\WINDOWS\System32\ADVAPI32.dll
ModLoad: 00007fff`69120000 00007fff`6917b000   C:\WINDOWS\System32\sechost.dll
ModLoad: 00007fff`69780000 00007fff`6989f000   C:\WINDOWS\System32\RPCRT4.dll
ModLoad: 00000000`65b80000 00000000`65bde000   D:\LIB\dlls\libquadmath_64-0.dll
ModLoad: 00000000`69800000 00000000`69823000   D:\LIB\dlls\libgcc_s_seh_64-1.dll
(2c9c.61a4): Break instruction exception - code 80000003 (first chance)
ntdll!LdrpDoDebuggerBreak+0x30:
00007fff`6a212ebc cc              int     3
0:000> g
ModLoad: 00007fff`69b90000 00007fff`69bbd000   C:\WINDOWS\System32\IMM32.DLL
ModLoad: 00007fff`69bd0000 00007fff`69ed8000   C:\WINDOWS\System32\combase.dll
ModLoad: 00007fff`67490000 00007fff`67502000   C:\WINDOWS\System32\bcryptPrimitives.dll
(2c9c.61a4): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for D:\LIB\dlls\libgfortran_64-3.dll - 
libgfortran_64_3!gfortran_f2c_specific__conjg_8+0x14ee0:
00000000`63218ae0 488b31          mov     rsi,qword ptr [rcx] ds:00000000`2e831ae0=????????????????

**Comment by martin-frbg** [comment]
Some quick comments:
- NO_CBLAS=1 sets NO_LAPACK=1 in Makefile.system
- LAPACK is _written_ in FORTRAN (and is just a copy of the "netlib" reference implementation, with only a handful of functions overloaded with optimized C equivalents), so building LAPACK induces a dependency on libgfortran, which in turn requires libquadmath by design if you use gfortran.
- I think there have been some problem reports from users of the "tdm" implementation of gcc in the past
- Are gcc and gfortran in your installation the same version and from the same source ? There may be incompatibilities between different versions
- There are a few recent unresolved issues about interactions between mingw and MSVC

**Comment by yyy-git** [comment]
@martin-frbg  Thank you for your reply. It's the netlib lapack requires Fortran, and it does be compiled rather than being a reference manual. 

MinGW does better, but in more complicated situation the program will crash too.

So the most thorough approach is to rewrite the LAPACK in C?(seems impossible...)

gcc (tdm64-1) 5.1.0
GNU Fortran (tdm64-1) 5.1.0

**Comment by martin-frbg** [comment]
There used to be a C "implementation" of LAPACK years ago, but it was not a rewrite but something created by running the f2c translator on the Fortran source (which means it ended up with some hard to read C code and a dependency on libf2c). I think a more promising development is the availability of the clang and flang compilers on Windows, but this is still a bit experimental (see the wiki)

**Comment by yyy-git** [comment]
so there is no easy way to solve this problem... 
Luckily what I need is just LAPACKE_dposv(), I can do it myself

**Comment by martin-frbg** [comment]
Maybe this can be revisited once #1421 is understood. Which version of Visual Studio are you using by the way?

**Comment by yyy-git** [comment]
I have tried /DYNAMICBASE:NO in visual stuido project, it works. But in mex(MATLAB C interface) it crashes. make DYNAMIC_ARCH=1 dosen't work. Sometimes openblas works in MATLAB but not in a C-program, and sometimes it's the opposite.All these confuse me, and when Fortran is removed, everything goes well.

Because my program have crashed numerous times with so many different settings, and occasionally the program is loading  an unexpected version of libopenblas(that's one of the reasons why I use LoadLibrary(),GetProcAddress() at last), I can' t make sure the information above is correct.

I am using visual stuido 2017.

**Comment by shoshia** [comment]
@yyy-git  are you trying win32 project in msvc or x64?

**Comment by yyy-git** [comment]
@shoshia  x64, some  BINARY=64 are omitted. If a win32 project is loading a x64 dll, the Error code and error string will tell me.This happened once.

**Comment by shoshia** [comment]
@yyy-git  Yes..
For me dynamicbase =no helped for lapacke  function calls. blas calls worked fine in both cases.
and build with dynamic_arch=1 was fine for both cases (blas, lapacke calls with and without dynamicbase=on )
never tryed mex.

**Comment by brada4** [comment]
#1421 already reports problem with LAPACK failing with DYNAMIC_ARCH absent.
MEX does not support just any compiler.
Probably easiest test for fresh build openblas DLL is to set BLAS_VERSION=libopenblas.dll LAPACK_VERSION=libopenblas.dll and then copy metered loop from octave benchmarks present in OpenBLAS source code.
Once test works you can use MEX module built with mex, that will link to any BLAS configured into matlab.
If you link openblas DLL into MEX file it is very likely to conflict with MKL that is included with MatLAB.

**Comment by yyy-git** [comment]
@brada4  But it can't explain the failure of LoadLibraray().I call LoadLibrary() with an explicit absolute path.My msvc project only calls something like cblas_daxpy in cblas, while the mex file calls cblas_daxpy and LAPACKE_dposv. Because  mkl.dll in MATLAB lacks C interface(cblas_), if  the reason is there are different functions with a same name, the function must be in lapack/lapacke.

No I am tired of these things.I'm going to wirte a LAPACKE_dposv myself, because I have to finish my algorithm asap.

**Comment by brada4** [comment]
Is it the same or better gfortran redist that was used to build OpenBLAS?
Does binary build of OpenBLAs (from sf.net) work
 - where you think you need it?
 - as matlab _VERSION?

if you have only fortran MKL DLLs in matlab you can still include cblas and lapacke wrappers in your module, or call fortran API from C directly. You dont need OpenBLAS in position of MEX module, you just need C wrapper for provided fortran lib.

**Comment by yyy-git** [comment]
All the binary builds I have found are compiled with TARGET=NEHALEM. They perform poorly.

I  have to make sure the blas/lapack is running in single-threaded mode, and all these should be done inside the MEX. So a function like openblas_set_num_threads()  is needed, but it's absent from blas.h or lapack.h provided by MATLAB.
 
MKL is unfriendly to AMD cpu. 
The program is running on AMD cpu.

matlab 2017a.

There is NO better choice.  I won't expain what I'm doing any more.

**Comment by brada4** [comment]
https://sourceforge.net/projects/openblas/files/v0.2.19/ both are with DYNAMIC_ARCH=1 and should adapt to running CPU. Maybe you need OPENBLAS_CORETYPE=Haswell to force on Broadwell or Zen series.
As I said - first integrate OpenBLAS in Matlab if you intend to support AMD, 
independently from
Making your MEX link to default (not C-) BLAS  and LAPACK(not -E) in MatlAB 
i.e Intel users go as nothing happened, AMD improve their matlab thanks to you.

**Comment by yyy-git** [comment]
sorry for my rudeness.

The binary build v0.2.19 chooses the Prescott branch. After setting OPENBLAS_CORETYPE=HASWELL ,it does the right thing.
Ocatave 4.2.1(windows) suffers from the same problem(openblas inside).It only runs at full speed when OPENBALS_CORETYPE,OPENBLAS_NUM_THREADS is set correctly.

The fundamental cause is that cpu dispatcher( a term from intel) of OpenBLAS makes rather  conventional choice when meeting unknown micro architectures. So after it's been compiled, the only way to change its mind is to set OPENBLAS_CORETYPE manually.  I don't think this is a good idea, because many users dosen't known even what CPU they are using, and all the trivial work shoud be done silently. But it seems environment variable is the only way to send .dll a message.
Anyway, this is another issue.

Agner Fog has studied the cpu dispatcher in intel's product like MKL, and he finds no perfect solution exists. Now I'm getting away from icc and mkl. 
There is a file named blas.spec in dir:matlab\r2017a\bin\win64, hours ago I tried to edit it. It works!!!AMD YES.
What you should do is to replace the "mkl.dll" in the second line with "libopenblas.dll",don't forget OPENBALS_CORETYPE,OPENBLAS_NUM_THREADS. It will surprise you if you are working on matlab+mkl+amd before.
blas.spec:
GenuineIntel Family * Model  * mkl.dll     # Intel processors
AuthenticAMD Family * Model  * mkl.dll     # AMD
"setenv BLAS_VERSION mkl.so:mklcompat.so for linux" is found in matlab's help, but the result is quite strange on my windows.
yet another issue...

My MEX has to use a separated blas because I want to control the version,num_threads,etc... precisely without altering the behavior of MATLAB itself.

**Comment by brada4** [comment]
Can you share your cpuid, like cpu-z screenshot? Let's have a look if it is supported in current development branch. I promise to add FAQ entry on how to build proper binary using linux from either 0.2.20 or develop - whichever supports your CPU right. It will take time to replace all default prescott with more common atom seen today.

**Comment by brada4** [comment]
Also if you get stable 10% speedup with different thread numbers than 
 it is worth integrating into openblas for everybody's good. I know some Matlab competitors who have documentation pieces saying to try less than all cores if you don't get linear speedup....

**Comment by martin-frbg** [comment]
So do we consider the original problem fixed (in which case it was probably a problem of tbm-gcc), or are we going off on a tangent here ? 
0.2.19 would be old enough to know neither Kaby Lake nor Ryzen, sadly there have not been "official" windows builds since then. (I guess I could create one, cross-compiled on Linux like I think xianyi's builds were, but I do not have write access to the project download page on sf.net as far as I know)

**Comment by brada4** [comment]
@martin-frbg such build would be very beneficial (sad that CI platforms keep artifacts hostage)
AMD being mentioned I suspect 0.2.20 will not reach ZEN (thats the need for screenshot)

**Comment by yyy-git** [comment]
I only tested dgemm()/cblas_dgemm(),matrix::A(4096x4096),matrix::B(4096x4096). A*B. 
Threadripper 1950x, 16 cores. SMT(AMD's hyper threadding) disabled.
Matlab(MKL): 16 threads, about 1.0 second.
Openblas: 16 threads, about 0.5 second. (0.4 second in High performacne mode(BIOS))
It has nothing to do about the optimization for AMD cpu. MKL is using SSE intrinsics on Threadripper 1950x, though it does support AVX, AVX2. It's the result of commercial competition. For further discussion, see Agner Fog's website. 

e.. I'm using two computers, the results above are  from KabyLake,i5-7500, expcept the  blas.spec and dgemm() test.  So the 0.2.19 build treats  KabyLake as Prescott.

It will be wonderful if there is a build binary for windows.

**Comment by brada4** [comment]
Can you post CPUID bits in place of marketing spreadsheet? Or should I look for it in some website you mention?
According to you cblas_dgemm was not available in Matlab MKL, so how you got to it?

**Comment by yyy-git** [comment]
@brada4 
I tested dgemm in MKL via matlab interface.
n = 4096;A=rand(n,n);B=rand(n,n);f=@()A*B;timeit(f)
From cpu usage in Task Manager I can see all cores are at full load.
The overhead is neglectable. Function timeit calls the specified function f multiple times, and computes the median of the measurements. I repeated the test many times and the result kept the same. In a word, the result is reliable.

This is the person who has stuied intel's cpu dispatcher.
[http://www.agner.org/optimize/blog/read.php?i=49](url)

**Comment by brada4** [comment]
I fail to see what it has to do with your Matlab dilemma. Can you help with model numbers of your CPU from cpu-z main page to check if current code identifies it correctly?

**Comment by martin-frbg** [comment]
Both Kaby Lake and current Ryzen should be detected by the current "develop" snapshot. Attached is a cross-compiled openblas.dll for testing - please rename from the .txt imposed by the issue tracker to openblasdll.zip and unpack.
[openblasdll.txt](https://github.com/xianyi/OpenBLAS/files/1680868/openblasdll.txt)

**Comment by yyy-git** [comment]
@martin-frbg  Thank you very much. Function openblas_get_corename() returns "Haswell".

@brada4  I have replaced the MKL in matlab with OpenBLAS on my AMD(1950x) machine.
the test script is simple:
 n = 4096;A=rand(n,n);B=rand(n,n);f=@()A*B;timeit(f)
OpenBLAS takes 0.5 second while MKL takes 1 second.

so, is cross-compile the solution?

rename .txt to .png. 
[cpuz_i5-7500.txt](https://github.com/xianyi/OpenBLAS/files/1681168/cpuz_i5-7500.txt)

**Comment by martin-frbg** [comment]
Haswell is newest we have, perhaps some tweaking for Threadripper would be desirable (lower AVX2 performance than Intel cpus I think ?). Cross-compile is probably not the only solution, it could be that your problem came from some incompatibility in the tdm build of gcc. None of the current developers seems to be using Windows much if at all, so interoperability with the various versions of VisualStudio is a bit of a gray area.

**Comment by brada4** [comment]
@yyy-git mind to add Matlab travails, especially basic integration, to FAQ? 
I had just basic knowledge on how to get gpgpu into Matlab without toolbox...

**Comment by yyy-git** [comment]
@brada4 
I won't have access to the desktop computers until March.
configure file:blas.spec decides which blas(dll) to load in MATLAB.
I have seen some MATLAB users complain about the speed of AMD cpu.  OpenBLAS will bring you  significantly increase of performance. But at your own risk!

Some problems I have met:
1. Once immediately after startup,I started MATLAB.Then I found  openblas in matlab running in single-threaded mode, though env-variable OPENBALS_NUM_THREAD was set to 16. I restarted the MATLAB, and  this time everything was right. 
2. When executing mex, the mex(a dll fundamentally) tends to look for dll in MATLAB dir first, rather than the current working dir.This may lead to a crash of MATLAB if you are using openblas in mex(I have numerous crashes these days). LoadLibrary()+GetProcAddress() is much safer, and I think that's exactly how MATLAB works. If you call openblas_set_num_threads() in MEX, then the number of openblas threads in MATLAB will be changed too. A simple workaround is to rename the "libopenblas.dll" in MATLAB dir.
